flax/linen/module.py

```diff
@@ -102,8 +104,11 @@ def module_stack(self):
     if not hasattr(self._thread_data, 'module_stack'):
       self._thread_data.module_stack = [None,]
     return self._thread_data.module_stack
+  @property
+  def at_top_level(self):
+    return len(self.module_stack) == 1
```

Comment:
This kind of check is pretty dangerous. It means there is a global "top-level". So the things that depend on at_top_level will never work if you do something like module.apply inside another module. The old flax also special cased the top frame and that resulted in some nasty edge cases for users.

Reply:
What exactly are we supposed to do with Modules passed in as attributes to an inner `module.apply` (in which case they would never naturally be parent=None)?  That's actually a very interesting edge-case to consider... I'm guessing we re-parent them in the same way we're talking about here?  Throw an error?  Perhaps we need to keep track of a module's "root" as well as its parent to even detect this kind of thing? (root might be useful for other stuff anyway...)

Reply:
Actually I'm being dumb - I'm pretty sure we already handle passing in submodules as attributes to internal `module.init/apply` just fine.  Some quick experimentation verifies that transforms, mutability, etc. work naturally when passing into internal `init/apply`.  It's acts like a trivial transform, there's absolutely nothing special here.

Which brings us back to the "top_level" check.  I think in this case there really -is- a unique top-top-level edge-case that we need to handle.  We don't naturally have `parent=None` modules appearing inside a module, this is very much only a top-top-level concern and _should_ be a no-op otherwise.

So I _generally_ agree with you that having top-level checks is sketchy and undesirable, but I think this is a special case where there really is a unique concern that only ever occurs at the "true" top level.